File: chess99-ios/Chess99/Presentation/Common/SoundManager.swift
import AVFoundation

/// Manages chess game sound effects.
/// Uses AVAudioPlayer for reliable playback.
///
/// Sound triggers match web frontend:
/// - move: Every legal piece movement
/// - capture: When a piece is captured
/// - check: When a move results in check
/// - gameEnd: When game completes (checkmate, stalemate, resignation)
final class SoundManager {
    static let shared = SoundManager()

    private var movePlayer: AVAudioPlayer?
    private var capturePlayer: AVAudioPlayer?
    private var checkPlayer: AVAudioPlayer?
    private var gameEndPlayer: AVAudioPlayer?
    private var isMuted = false

    private init() {
        // Configure audio session for game sounds
        try? AVAudioSession.sharedInstance().setCategory(.ambient, mode: .default)

        // Load sound files from bundle
        // NOTE: Add these sound files to the Xcode project:
        // movePlayer = loadSound("move", ext: "mp3")
        // capturePlayer = loadSound("capture", ext: "mp3")
        // checkPlayer = loadSound("check", ext: "mp3")
        // gameEndPlayer = loadSound("game_end", ext: "mp3")
    }

    func playMove() { play(movePlayer) }
    func playCapture() { play(capturePlayer) }
    func playCheck() { play(checkPlayer) }
    func playGameEnd() { play(gameEndPlayer) }

    func setMuted(_ muted: Bool) { isMuted = muted }

    private func play(_ player: AVAudioPlayer?) {
        guard !isMuted, let player = player else { return }
        player.currentTime = 0
        player.play()
    }

    private func loadSound(_ name: String, ext: String) -> AVAudioPlayer? {
        guard let url = Bundle.main.url(forResource: name, withExtension: ext) else { return nil }
        return try? AVAudioPlayer(contentsOf: url)
    }
}

File: chess99-android/app/src/main/java/com/chess99/presentation/common/SoundManager.kt
package com.chess99.presentation.common

import android.content.Context
import android.media.AudioAttributes
import android.media.SoundPool
import com.chess99.R
import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Manages chess game sound effects.
 * Uses SoundPool for low-latency playback.
 *
 * Sound triggers match web frontend:
 * - move: Every legal piece movement
 * - check: When a move results in check
 * - gameEnd: When game completes (checkmate, stalemate, resignation)
 */
@Singleton
class SoundManager @Inject constructor(
    @ApplicationContext private val context: Context,
) {
    private val soundPool: SoundPool
    private var moveSound: Int = 0
    private var checkSound: Int = 0
    private var gameEndSound: Int = 0
    private var captureSound: Int = 0
    private val loadedSounds = java.util.concurrent.ConcurrentHashMap.newKeySet<Int>()
    private var isMuted = false

    init {
        val attrs = AudioAttributes.Builder()
            .setUsage(AudioAttributes.USAGE_GAME)
            .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
            .build()

        soundPool = SoundPool.Builder()
            .setMaxStreams(4)
            .setAudioAttributes(attrs)
            .build()

        soundPool.setOnLoadCompleteListener { _, sampleId, status ->
            if (status == 0) loadedSounds.add(sampleId)
        }

        // Reuse Chess99's existing web game sounds in the native package.
        moveSound = soundPool.load(context, R.raw.move, 1)
        checkSound = soundPool.load(context, R.raw.check, 1)
        gameEndSound = soundPool.load(context, R.raw.game_end, 1)
        captureSound = soundPool.load(context, R.raw.capture, 1)
    }

    fun playMove() { play(moveSound) }
    fun playCheck() { play(checkSound) }
    fun playGameEnd() { play(gameEndSound) }
    fun playCapture() { play(captureSound) }

    fun setMuted(muted: Boolean) { isMuted = muted }

    private fun play(soundId: Int) {
        if (isMuted || !loadedSounds.contains(soundId)) return
        soundPool.play(soundId, 1.0f, 1.0f, 1, 0, 1.0f)
    }

    fun release() {
        soundPool.release()
    }
}
